### PR TITLE
Dynamic programming algorithm for computing distances and geodesics for the SRV metric

### PR DESCRIPTION
**Pull request by Abdellaoui-Souhail**
## Checklist

- [ ] My pull request has a clear and explanatory title.
- [ ] If necessary, my code is [vectorized](https://www.geeksforgeeks.org/vectorization-in-python/).
- [ ] I added appropriate unit tests.
- [ ] I made sure the code passes all unit tests. (refer to comment below)
- [ ] My PR follows [PEP8](https://peps.python.org/pep-0008/) guidelines. (refer to comment below)
- [ ] My PR follows [geomstats coding style](https://github.com/geomstats/geomstats/blob/master/docs/contributing.rst#coding-style-guidelines) and API.
- [ ] My code is properly documented and I made sure the documentation renders properly. ([Link](https://github.com/geomstats/geomstats/blob/master/docs/contributing.rst#documentation))
- [ ] I linked to issues and PRs that are relevant to this PR.

### CODE REVIEW

File: geomstats/geometry/discrete_curves.py [hunk 1]
```diff
@@ -1936,6 +1987,265 @@ def horizontal_path(t):
...
+            velocity = k_sampling_point * (point[1:, :] - point[:-1, :])
+            square_root_velocity = np.sqrt(np.sum(gs.abs(velocity), axis=-1))
+            _q = np.array([i / j for (i, j) in zip(velocity, square_root_velocity)])
+            q = np.array([_q[int(np.floor(i * (k_sampling_point / n)))]
+                          for i in range(n)])
+
+            return q
+
+        def reparametrisation_q(q, gamma):
+
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2058]: Missing docstring. What does this do ?

File: geomstats/geometry/discrete_curves.py [hunk 2]
```diff
@@ -1936,6 +1987,265 @@ def horizontal_path(t):
...
+                raise AssertionError(
+                    "The square root velocity framework "
+                    "is only defined for discrete curves "
+                    "with distinct consecutive sample points."
+                )
+            k_sampling_point = point.shape[-2] - 1
+            velocity = k_sampling_point * (point[1:, :] - point[:-1, :])
+            square_root_velocity = np.sqrt(np.sum(gs.abs(velocity), axis=-1))
+            _q = np.array([i / j for (i, j) in zip(velocity, square_root_velocity)])
+            q = np.array([_q[int(np.floor(i * (k_sampling_point / n)))]
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2052]: All the `np.` should be replaced by `gs.`

File: geomstats/geometry/discrete_curves.py [hunk 3]
```diff
@@ -1936,6 +1987,265 @@ def horizontal_path(t):
...
+        def compute_integral_restricted(q_1, q_2, x_min, x_max, y_min, y_max):
+            """Compute a value of an integral over a subinterval.
+
+            Compute n * the value of the integral of
+            t --> q_1(t)*q_2(gamma(t))*|gamma(t)|^(1/2) over [x_min,x_max]
+            where gamma restricted to [x_min,xmax] is the linear
+            interpolation between y_min et y_max.
+            gamma(x_min) = y_min and gamma(x_max) = y_max
+            """
+            # Gamma's parameters
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2084]: No inline comment.

File: geomstats/geometry/discrete_curves.py [hunk 4]
```diff
@@ -1936,6 +1987,265 @@ def horizontal_path(t):
...
+
+            """
+            At each step we compute value_1 and value_2 which are related to
+            the value of the integral restricted to [lower_bound,upper_bound]:
+            value = gamma_slope^(1/2) ( value_1*gamma_slope + value_2) / n
+            We can compute value_1 and value_2 in closed form, we have introduced
+            them to make the least amount of operation : the multiplications by
+            gamma_slope, gamma_slope^(1/2) and the division by n is done at the
+            very end instead of doing it at every interval.
+            """
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2105]: Docstring should only be at the beginning of the method, not in the middle.

File: geomstats/geometry/discrete_curves.py [hunk 5]
```diff
@@ -1936,6 +1987,265 @@ def horizontal_path(t):
...
+                    j += 1
+                lower_bound = upper_bound
+
+            value = math.pow(gamma_slope, 1 / 2) * \
+                ((gamma_slope * value_1) / 3 + value_2 / 2 + value_3)
+
+            return value
+
+        def compute_integral_restricted_constant(
+                q_1, q_2, x_min, x_max, y_min, y_max):
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2147]: What is the difference between this function and the previous one, and do both really need to be implemented here ? If there is one that is better than the other, the other should be removed.

**Review by alebrigant — CHANGES_REQUESTED** [review]
Here are a few first comments. I haven't looked at the tests yet, because most of them don't pass. 

An important issue is the organization of the code, which will be affected by the modifications made in https://github.com/luisfpereira/geomstats/pull/8#issuecomment-1522004958. @luisfpereira, should we maybe wait for your PR to pass before continuing on this ? This is a PR that implements another method to compute geodesics in the quotient space, when the total metric is the `SRVMetric`.

**Comment by luisfpereira** [comment]
@alebrigant, I think it is better to wait for the merge. It will not be a long wait though (I'm probably finishing it today or very beginning of next week).

**Comment by alebrigant** [comment]
Ok makes sense. @Abdellaoui-Souhail I think it's still worth it, in the meantime, to address previous comments and work on the tests. This means less things to do later !

File: geomstats/geometry/discrete_curves.py [hunk 6]
```diff
@@ -1842,6 +1843,244 @@ def horizontal_path(t):
...
+        between initial_curve and end_curve reparametrized by gamma, and output
+        the corresponding distance, using a dynamic programming algorithm.
+
+        The objective can be expressed in terms of square root velocity (SRV)
+        representations: it is equivalent to finding the gamma that maximizes
+        the L2 scalar product between initial_srv and end_srv@gamma where initial_srv
+        is the SRV representation of the initial curve and end_srv@gamma is the SRV
+        representation of the end curve reparametrized by gamma, i.e
+        .. math::
+        end_srv@\gamma(t) = end_srv(\gamma(t))\cdot\lvert\gamma(t)\rvert^\frac{1}{2}
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:1860]: Maybe replace "\lvert\gamma(t)\rvert" by "| \gamma(t) |" which gives the same output in Latex but is more readable in the docstring ?

File: geomstats/geometry/discrete_curves.py [hunk 7]
```diff
@@ -1842,6 +1843,244 @@ def horizontal_path(t):
...
+
+        The dynamic programming algorithm assumes that for every subinterval
+        :math: '\left[\frac{i}{n},\frac{i+1}{n}\right]' of :math: '\left[0,1\right]',
+        gamma is linear.
+
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:1865]: Could you add a reference to a paper where dynamic programming is explained ? Like it is done e.g. [here](https://github.com/geomstats/geomstats/blob/master/geomstats/geometry/discrete_curves.py#LL816C5-L816C15).

File: geomstats/geometry/discrete_curves.py [hunk 8]
```diff
@@ -1842,6 +1843,244 @@ def horizontal_path(t):
...
+        end_curve_reparametrized = reparametrization_curve(
+            end_curve, gamma[(n_discretization, n_discretization)])
+
+        geodesic = self.total_space.metric.geodesic(
+            initial_curve, end_curve_reparametrized)
+
+        return {"geodesic": geodesic, "distance": distance}
+
     def align(self, point, base_point, n_times=20, threshold=1e-3):
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2084]: Now we have 2 methods to align two curves and compute horizontal geodesics: (1) iterative horizontal projection and (2) dynamic programming. We should be able to call either of these methods in `align`, as well as in `horizontal_geodesic`. So what I suggest is:
- Rename `horizontal_geodesic` to `_iterative_horizontal_projection`
- Create a new `horizontal_geodesic` that calls either `_dynamic_programming` or `_iterative_horizontal_projection`
- Add a `method` variable to `align` and call `horizontal_geodesic(method=method)`.

File: geomstats/geometry/discrete_curves.py [hunk 9]
```diff
@@ -1880,19 +2119,47 @@ class SRVQuotientMetric(QuotientMetric):
     reparametrization corresponds to resampling.
     """
 
-    def geodesic(self, initial_point, end_point, threshold=1e-3):
+    def geodesic(self, initial_point, end_point, method="horizontal projection",
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2122]: "iterative horizontal projection"

File: geomstats/geometry/discrete_curves.py [hunk 10]
```diff
@@ -1880,19 +2119,47 @@ class SRVQuotientMetric(QuotientMetric):
...
-        return self.fiber_bundle.horizontal_geodesic(
-            initial_point, end_point, threshold
-        )
+        if method == "horizontal projection" :
+            return self.fiber_bundle.horizontal_geodesic(
+                initial_point, end_point, threshold
+            )
 
-    def dist(self, point_a, point_b, n_times=20, threshold=1e-3):
+        elif method == "dynamic programming" :
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2136]: `elif` is not necessary here: see DeepSource errors.

File: geomstats/geometry/discrete_curves.py [hunk 11]
```diff
@@ -1880,19 +2119,47 @@ class SRVQuotientMetric(QuotientMetric):
...
 
         The geodesics between unparametrized curves for the quotient metric are
         projections of the horizontal geodesics in the total space of parameterized
         curves. Since in practice shapes can only be encoded by parametrized curves,
         geodesics are given in the total space.
         """
-        return self.fiber_bundle.horizontal_geodesic(
-            initial_point, end_point, threshold
-        )
+        if method == "horizontal projection" :
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2131]: "iterative horizontal projection"

File: geomstats/geometry/discrete_curves.py [hunk 12]
```diff
@@ -1880,19 +2119,47 @@ class SRVQuotientMetric(QuotientMetric):
...
+            )
 
-    def dist(self, point_a, point_b, n_times=20, threshold=1e-3):
+        elif method == "dynamic programming" :
+            point_ndim = gs.ndim(initial_point)
+            if point_ndim == 2:
+                dp = self.fiber_bundle._dynamic_programming(
+                    initial_point, end_point, n_discretization, max_slope)
+                return dp["geodesic"]
+            else :
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2142]: Same as above.

File: geomstats/geometry/discrete_curves.py [hunk 13]
```diff
@@ -1842,6 +1843,244 @@ def horizontal_path(t):
...
+            k_sampling_point = point.shape[-2] - 1
+            velocity = k_sampling_point * (point[1:, :] - point[:-1, :])
+            square_root_velocity = gs.sqrt(gs.sum(gs.abs(velocity), axis=-1))
+            _srv = gs.array([i / j for (i, j) in zip(velocity, square_root_velocity)])
+            srv = gs.array([_srv[int(gs.floor(i * (k_sampling_point
+                            / n_discretization)))] for i in range(n_discretization)])
+
+            return srv
+
+        def reparametrization_curve(curve, gamma):
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:1921]: Maybe change to just `reparametrize(curve, gamma)` ?

File: geomstats/geometry/discrete_curves.py [hunk 14]
```diff
@@ -1842,6 +1843,244 @@ def horizontal_path(t):
...
+            """Compute end_curve reparametrized by gamma.
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:1922]: Reparametrize curve by gamma ?

File: geomstats/geometry/discrete_curves.py [hunk 15]
```diff
@@ -1880,19 +2119,47 @@ class SRVQuotientMetric(QuotientMetric):
...
+                n_points = initial_point.shape[0]
+                results = gs.zeros(n_points, dtype=object)
+
+                for i in range(n_points):
+                    dp = self.fiber_bundle._dynamic_programming(
+                        initial_point[i], end_point[i], n_discretization, max_slope
+                    )
+                    results[i] = dp["geodesic"]
+
+                return results
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2154]: All of this code should be moved to `horizontal_geodesic`.

File: geomstats/geometry/discrete_curves.py [hunk 16]
```diff
@@ -1907,28 +2174,63 @@ def dist(self, point_a, point_b, n_times=20, threshold=1e-3):
             Discrete curve.
         point_b : array-like, shape=[k_sampling_points, ambient_dim]
             Discrete curve.
+        method : str, {"horizontal projection", "dynamic programming"}
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2177]: "iterative horizontal projection"

**Review by alebrigant — CHANGES_REQUESTED** [review]
Almost there ! Still a few comments.

File: geomstats/geometry/discrete_curves.py [hunk 17]
```diff
@@ -1842,7 +1844,326 @@ def horizontal_path(t):
...
+        Outputs
+        -------
+        geodesic: callable
+            Time parametrized geodesic.
+        dist : float
+            Quotient distance between intial_curve and end_curve.
+
+        References
+        ----------
+        Mio, Washington & Srivastava, Anuj & Joshi, Shantanu. (2007).
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:1890]: Thanks for adding the reference. Could you just copy the style used in geomstats ? See [here](https://github.com/geomstats/geomstats/blob/master/geomstats/geometry/discrete_curves.py#LL816C5-L816C15).

File: geomstats/geometry/discrete_curves.py [hunk 18]
```diff
@@ -1907,28 +2231,48 @@ def dist(self, point_a, point_b, n_times=20, threshold=1e-3):
...
+            return gs.sum(velocity_norms) / n_times
+
+        if method == "dynamic programming" :
+            results = self.fiber_bundle._dynamic_programming(
+                point_a, point_b, n_discretization, max_slope)
+            return results["distances"]
+
+        else :
+            raise AssertionError(
+                "Method not implemented"
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2277]: Maybe be more specific here: "The only methods implemented are iterative horizontal projection and dynamic programming."

File: geomstats/geometry/discrete_curves.py [hunk 19]
```diff
@@ -1907,28 +2231,48 @@ def dist(self, point_a, point_b, n_times=20, threshold=1e-3):
...
+                horizontal_geod_velocity, horizontal_geod[:-1]
+            )
+            return gs.sum(velocity_norms) / n_times
+
+        if method == "dynamic programming" :
+            results = self.fiber_bundle._dynamic_programming(
+                point_a, point_b, n_discretization, max_slope)
+            return results["distances"]
+
+        else :
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2275]: Unnecessary else after return.

File: geomstats/geometry/discrete_curves.py [hunk 20]
```diff
@@ -1842,7 +1844,326 @@ def horizontal_path(t):
...
+        if method == "iterative horizontal projection" :
+            return self._iterative_horizontal_projection(
+                initial_point, end_point, threshold)
+
+        if method == "dynamic programming" :
+            results = self._dynamic_programming(
+                initial_point, end_point, n_discretization, max_slope)
+            return results["geodesics"]
+
+        else :
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2160]: Unnecessary else after return.

File: geomstats/geometry/discrete_curves.py [hunk 21]
```diff
@@ -1842,7 +1844,326 @@ def horizontal_path(t):
...
+        point_ndim = gs.ndim(initial_curve)
+        results = {"geodesics" : [], "distances" : []}
+
+        if point_ndim == 2:
+            dp = self._dynamic_programming_single(
+                initial_curve, end_curve, n_discretization, max_slope)
+            results["geodesics"] = dp["geodesic"]
+            results["distances"] = dp["distance"]
+            return results
+        else :
```
Comment (alebrigant) [inline geomstats/geometry/discrete_curves.py:2123]: Unnecessary else after return.

**Review by alebrigant — CHANGES_REQUESTED** [review]
Thanks for the changes ! Could you address the final minor comments and then we can merge !

**Pull request merged**